File: src/harness/error_policy.py
from __future__ import annotations

from .types import ToolExecutionResult


class ErrorPolicy:
    def __init__(self, tool_retry_limit: int = 0) -> None:
        self.tool_retry_limit = max(0, tool_retry_limit)

    def should_retry_tool(self, result: ToolExecutionResult, attempt: int) -> bool:
        return (not result.ok) and result.retryable and attempt <= self.tool_retry_limit

File: ts/src/error-policy.ts
/**
 * Error policy — determines whether a failed tool call should be retried.
 */

import type { ToolExecutionResult } from "./types.js";

export class ErrorPolicy {
  public readonly toolRetryLimit: number;

  constructor(toolRetryLimit = 0) {
    this.toolRetryLimit = Math.max(0, toolRetryLimit);
  }

  /** Returns true if the tool should be retried. */
  shouldRetryTool(result: ToolExecutionResult, attempt: number): boolean {
    return !result.ok && result.retryable && attempt <= this.toolRetryLimit;
  }
}
